### ISSUE TITLE
Prevent "prisma migrate dev" from being possible to run on a production database.

### ISSUE THREAD

**Issue by tombburnell** [prisma/prisma#16587]
## Problem
If you run `prisma migrate dev` on a production database it attempt to run and if there is any migrations with changed data it suggests resetting it which you could do by accident. Frightening stuff!

## Suggested solution
Detect if it is a production database (maybe lack of shadow db) and exit immediately with big warning.

## Additionally..
I'd also like to see `prisma migrate deploy` provide a list of all SQL statements that will be executed with any that could delete data highlighted.

**Comment by tombburnell** [comment]
Any thoughts?

**Comment by janpio** [comment]
This is now tracked as a potential improvement, and when we have time to look at this and consider it, we will.

Generally the "detect production" is the biggest problem here. There is no real, reliable way that comes to mind how to do that.

**Comment by tombburnell** [comment]
Thanks. 
Does production have a shadow database? Could you use lack of shadow DB as an indicator?

**Comment by janpio** [comment]
Unfortunately not. Shadow databases only exist during the runtime of `migrate dev`, where they are created and then also destroyed.

**Comment by janpio** [comment]
We do have `_prisma_migrations`, but that would leave us with the problem that the user has to somehow flag the deployment as production and then use some command to write that. If they have to do this, there are probably easier options than to store that in a database (e.g. the presence of an ENV var or a file). 

The problem then is just, if that step is forgotten, the behavior is the same as right now. And expecting to be protected, but then not being is an easy way to make such a situation even worse.

**Comment by tombburnell** [comment]
Why does the user have flag it? the command is different for dev and prod.

when you run **prisma migrate dev** -> that should mark it as a dev db
when you run **prisma migrate deploy** -> that should mark is as prod db.

If you then try to run **prisma migrate dev** when it's been set to prod (after running **prisma migrate deploy**) then I think it should fail immediately.

**Comment by janpio** [comment]
I like this suggestion.

Unfortunately usage of `migrate deploy` does not indicate production. It is totally ok to get new migrations via a pull of your repository, and then apply these with `migrate deploy` while you already have changes in your schema that are pending and would appear in a migration via `migrate dev`.

Maybe there could another command for that though, some parameter for `dev`, that would then indeed allow this separation of responsibilities.

**Comment by tombburnell** [comment]
yeah I was going to ask whether migrate deploy is ever used on dev - since that would be the limiting factor.

I think it would be wise to split it as you suggest.

- Make **migrate dev** only work on dev
- Make **migrate deploy** only work on prod
- If you need to apply migrations to dev use something like **migrate dev --deploy**

thanks

**Comment by capaj** [comment]
@janpio 
one thing that could help for me, would be to write an amount of data that will be deleted.
You can easily miss a db url in the text, but if you see an output from `SELECT pg_size_pretty( pg_database_size('db_name') );`

that could help people realize they have the wrong DB when it shows several terabytes.

**Comment by acron0** [comment]
I think this would be a great feature and doesn't need to even be that complicated. It would be trivial to (manually) add an empty table like `_prisma_migrate_production_lock` and then have Prisma check for the presence of this table before performing `migrate` operations, and fail for anything other than `deploy` or `resolve`.

**Comment by qrcf** [comment]
Any updates on this?
